terraform/modules/application/kms.tf

```diff
@@ -0,0 +1,52 @@
+resource "aws_kms_key" "this" {
+  description         = "KMS key used to encrypt SSM parameters."
+  enable_key_rotation = true
+}
+
+resource "aws_kms_alias" "this" {
+  name          = "alias/${var.prefix}-app-ssm-parameters"
+  target_key_id = aws_kms_key.this.key_id
+}
+
+resource "aws_kms_key_policy" "this" {
+  key_id = aws_kms_key.this.id
+  policy = data.aws_iam_policy_document.kms.json
+}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_iam_policy_document" "kms" {
+  #checkov:skip=CKV_AWS_109:Only assigning the kms:GenerateDataKey and kms:Decrypt permissions led to a 'you won't be able to manage the key once created' error
+  #checkov:skip=CKV_AWS_111:Only assigning the kms:GenerateDataKey and kms:Decrypt permissions led to a 'you won't be able to manage the key once created' error
+  #checkov:skip=CKV_AWS_356:Only assigning the kms:GenerateDataKey and kms:Decrypt permissions led to a 'you won't be able to manage the key once created' error
+  statement {
+    principals {
+      type        = "AWS"
+      identifiers = [var.ecs_task_execution_role_arn]
+    }
+
+    actions = [
+      "kms:GenerateDataKey",
+      "kms:Decrypt"
+    ]
```

Comment:
Are both permissions needed here?

Reply:
Actually to be fair in this case the role will definitely need to decrypt, and I'm less bothered about whether the generate data key permission is technically needed - but it's still worth thinking about.

Reply:
I believe `kms:GenerateDataKey` is generally needed when you're doing client-side decryption / encryption, which I don't believe we are (the S3 buckets use `SSE-KMS` and some looking up on param store / secrets suggests it's server side here too though it isn't obvious)

I'll remove them and see if we break anything in terms of S3 / Param Store / Secrets for the app 😅

Reply:
Removed 👍 , over the weekend some export jobs will try to run which will give some ideas, and I'll redeploy the app to ECS to see if it can still pull down params / secrets

Reply:
I tested the theory out by removing the `kms:GenerateDataKey` everywhere, but it broke the Export job and Bulk Upload process. The rest of the app seemed to behave OK but I felt it was better to retain this action everywhere it was originally removed from to avoid quietly breaking something, balanced with fact this action shouldn't cause concern if it is there for being overly permissive 👍

Reply:
That's not surprising - I would expect `kms:GenerateDataKey` to be needed by the task role for the keys used for the bulk upload and export buckets, because the app puts things into those buckets using that role. I don't expect anything using the task execution role to putting data into ssm parameters though - I think this is a separate thing to look at.

Reply:
Removed the `kms:GenerateDataKey` from everywhere that isn't to do with S3 (which meant anything to do with KMS and the task execution role), redeployed the app (it was able to pull down secrets and parameters) and satisfied everything looks to be working fine without it 👍